### DISCUSSION TITLE
V4L2 settings not applied on boot, but applied after restart of Crowsnest from within Mainsail

### DISCUSSION THREAD

**Discussion by emailcraigs** [mainsail-crew/mainsail#2083] [General]
When I first boot Klipper and Mainsail comes up, the V4L2 settings are not applied. I can tell because my camera shows out of focus. If I restart Crowsnest from the Mainsail menu it applies the V4L2 settings and when I refresh the camera screen it shows up focused.

**Comment by mryel00** [comment]
Please add a log with `log_level: debug` and `delete_log: false`, where you reproduce the issue once and then fix it.
Also make sure that your Crowsnest is up to date this time, as we cannot give support for the outdated versions.

You might also want to check out the `-s` parameter, that got added with v4.1.0: https://github.com/mainsail-crew/crowsnest/issues/117#issuecomment-1806943132

**Reply by emailcraigs** [reply]
[crowsnest CTRL settings problem.log](https://github.com/user-attachments/files/18134101/crowsnest.CTRL.settings.problem.log)

I upgraded and set the -s parameter, but the problem persists. The log is a shutdown, restart with camera not in focus, and then I restart crowsnest from the mainsail power menu, and the camera refreshes and shows in focus.

The logs seem to show it applying the setting so I can only assume that on boot it is trying to set it before its ready? I think the -s was suppose to fix that?

**Reply by mryel00** [reply]
So it seems most likely to be a similar issue to this issue https://github.com/mainsail-crew/crowsnest/issues/96
We sadly cannot really do anything about it, as v4l2 is reporting wrong values. So we cannot even know, if the value got set or not.
In the issue someone did a small POC that you can try to integrate on your own, but I won't give support for that.
As I'm currently rewriting the whole codebase to python, I will not try to fix this in the bash version either. But I will try to fix it with this rewrite https://github.com/mainsail-crew/crowsnest/pull/270
In theory someone can already test it, but I don't think I already added a fix for the issue. Atm it's also a drop in replacement, but in future it will need a manual update, without breaking any old installations as with v3->v4.
If you feel confident enough to fix some minor problems on your own, I could let you know as soon as I implemented it in v5 that you can test it.

For people reading this, that are using a Pi with PiOS, you can try out camera-streamer. There should be a fix directly inside camera-streamer.

**Reply by emailcraigs** [reply]
I'm not super Unix knowledgeable. Is it possible to add somewhere a piece of code to automatically restart Crowsnest that would simulate the restart function on the power button of Mainsail? Basically that is what I do now. It starts and then I restart Crowsnest and it works perfectly until the next hard boot.

**Reply by mryel00** [reply]
I'm not that Unix knowledgeable either, so there might be easier options.
`systemctl restart crowsnest` would be a command to restart Crowsnest from the command line.
You can then create a bash script that's checking for a croswsnset process to start. Then wait like 10-15s that it started completely and restart it afterwards. To use it at startup you would need to make a service for it.
ChatGPT might be of good help for such a topic.

**Comment by emailcraigs** [comment]
With your advice and Between ChatGPT and Googlefi I got enough to piece it together. It now automatically restarts crowsnest and comes up focused right. Thanks!
